sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,658 @@
+
+#### Handling shadowed device global variables
+
+The example above shows a simple case where the user's device global variables
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+could reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a temporary variable that holds the address of the
+shadowed device global:
+
+```
+namespace {
+const void *__sycl_UNIQUE_STRING = &FuBar;  // References 'FuBar' in the
+                                            // unnamed namespace
+}
+
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::FuBar,
+    /* mangled name of '::FuBar' */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+  device_global_map::add(::__sycl_UNIQUE_STRING,
+    /* mangled name of '::(unnamed)::FuBar' with unique suffix appended */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+}
+
+} // namepsace (unnamed)
+} // namespace sycl::detail
+```
+
+The `__sycl_UNIQUE_STRING` variable is defined in the same namespace as the
+second `FuBar` device global, so it can reference the variable through
+unqualified name lookup.  Furthermore, the name of the temporary variable
+(`__sycl_UNIQUE_STRING`) is globally unique, so it is guaranteed not to be
+shadowed by any other name in the translation unit.  This problem with variable
+shadowing is also a problem for the integration footer we use for
+specialization constants.  See the [specialization constant design document][3]
+for more details on this topic.
+
+[3]: <SpecializationConstants.md>
+
+### Changes to the `sycl-post-link` tool
+
+The `sycl-post-link` tool performs its normal algorithm to identify the set of
+kernels and device functions that are bundled together into each module.  Once
+it identifies the functions in each module, it scans those functions looking
+for references to global variables of type `device_global`.  If any device
+global variable decorated with `sycl-device-global-image-life` appears in more
+than one module, the `sycl-post-link` tool issues an error diagnostic:
+
+```
+error: device_global variable <name> with property "device_image_scope"
+       is contained in more than one device image.
+```
```

Comment:
Just wondering, why we should emit error here?

My understanding is that even if `device_global` marked with `sycl-device-global-image-life`, it can be used within several device images with assumption that address of it is not shared between those device images. Like in each device image there are two kernels, which operate on that global, but no data about the global is shared between device images.

However, I guess it would be hard to prove that such usage of `device_global` is correct and therefore we simply ask our users not to do so, for the purpose of preventing errors, right?